pythonforandroid/python.py

```diff
@@ -249,26 +260,47 @@ def include_root(self, arch_name):
     def link_root(self, arch_name):
         return join(self.get_build_dir(arch_name), 'android-build')
 
+    def compile_python_files(self, dir):
+        '''
+        Compile the python files (recursively) for the python files inside
+        a given folder.
+
+        .. note:: python2 compiles the files into extension .pyo, but in
+            python3, and as of Python 3.5, the .pyo filename extension is no
+            longer used...uses .pyc (https://www.python.org/dev/peps/pep-0488)
+        '''
+        args = [self.ctx.hostpython]
+        if self.ctx.python_recipe.name == 'python3':
+            args += ['-OO', '-m', 'compileall', '-b', '-f', dir]
+        else:
+            args += ['-OO', '-m', 'compileall', '-f', dir]
+        subprocess.call(args)
```

Comment:
Why not using the `sh` module?

Reply:
Because it fails:

  - in python 3 fails when the compilation is about to finish (after compile zipfile module) without a clear message of why it failed...still seems that the files are compiled well but to bypass the error we would had to put a `try/except` (mmm...I don't like that without knowing why it fails...)
  - in python 2  the compilation not even starts and complains about elf32 ...

Then I realized that we have another call to compile all files in meth:`~pythonforandroid.bootstraps.common.build.build.compile_dir` and in there we used the `subprocess.call`...so I tried with that...and it works fine for both versions of python :smile:

:thinking:...my guess is that some of our environment variables is causing some kind of conflict between hostpython and python when we try to use the compileall command with sh...but I'm not sure about that...maybe @inclement, @tito, @hackalog or @JonasT  could give us some more light in here...

Reply:
Just to add as extra explanation: why you have to look up how `sh` works because it's unclear how it handles 1. the environment, 2. stdout/stderr buffering, 3. stdout/stderr redirect, and similar things. (for some of which p4a is using a really horrible wrapper in the logging file) So sure, the call itself is simple enough that it's not much of an issue, but if you care for details the `sh` module IMHO is really kind of a problem just because it isn't obvious for someone only familiar with `submodule` what it'll do or which parameters make it do certain things

Reply:
I'm not opposed to using subprocess, but I do think the sh module is pretty nice. I don't really see anything you mention as an issue, it isn't obvious how subprocess works either until you read the documentation, although it's fair to say that more people will have done so in advance. Most people will never have to read the details of either.

That said, sh is overused in p4a partly due to me being lazy when originally duplicating the functionality of the old-p4a shell scripts.

I would most like to use some nice pythonic file management module for a lot of things that in a shell script would be calls to cp, rm, mkdir etc.. It annoys me that `shutil` and `os` are both quite poorly structured. https://github.com/inclement/nosh is the kind of thing I had in mind.

Reply:
@inclement not sure I agree with `Most people will never have to read the details of either`, at a point where that starts to matter people will have, and that's exactly the point where `sh` then ends up problematic. Therefore I personally prefer to stick to core modules for such basic functionality even if they might arguably be not nearly as pythonic as they should be, but of course that's just my opinion

Reply:
@opacam OK failing is a fair enough point, I was just wondering anyway :stuck_out_tongue: 
@JonasT I didn't know the `sh` module before knowing p4a, I can't tell if it's that great, but I can tell it's not that bad. I'm also more a fan of going with core modules as much as possible. I was asking @opacam why he didn't use, because this is what's being used by the project so far, so not using it could have a reason I was curious about